### PR TITLE
use FormattedSymbol

### PR DESCRIPTION
**Pull request by heejaechang**
fix devdiv bug #118143

### CODE REVIEW

**Comment by heejaechang** [comment]
@AlekseyTs @mavasani @srivatsn @basoundr can you take a look?

**Comment by AlekseyTs** [comment]
@heejaechang Could you clarify the goal of the change. Is it to use FormattedSymbol everywhere or for specific diagnostics?

File: src/Compilers/VisualBasic/Portable/Errors/MessageProvider.vb [hunk 1]
```diff
@@ -102,12 +102,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             End If
 
             ' cases where we actually want fully qualified name
-            If errorCode = ERRID.ERR_AmbiguousAcrossInterfaces3 OrElse
-               errorCode = ERRID.ERR_TypeConflict6 OrElse
-               errorCode = ERRID.ERR_ExportedTypesConflict OrElse
-               errorCode = ERRID.ERR_ForwardedTypeConflictsWithDeclaration OrElse
-               errorCode = ERRID.ERR_ForwardedTypeConflictsWithExportedType OrElse
-               errorCode = ERRID.ERR_ForwardedTypesConflict Then
+            If errorCode = ERRID.ERR_TypeConflict6 Then
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Errors/MessageProvider.vb]: Why do we still have to keep this special case?

**Comment by heejaechang** [comment]
@AlekseyTs you opened this bug - devdiv #118143 

quite a long time ago though :)

...
Aleksey Tsingauz commented a year ago
FormattedSymbol/CustomSymbolDisplayFormatter should be used instead when DiagnosticInfo is created.
....

File: src/Compilers/VisualBasic/Portable/Errors/MessageProvider.vb (hunk 1 above)
Comment (heejaechang) [inline src/Compilers/VisualBasic/Portable/Errors/MessageProvider.vb]: 2 other cases was a bit more involved. doing here seems much simpler.

Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Errors/MessageProvider.vb]: I think we shouldn't have any special handling here.

**Comment by AlekseyTs** [comment]
@heejaechang I know I opened the bug, just trying to get clarity on your intent with this change. The description says: "use FormattedSymbol **when possible**" Feels too broad, hence the question.

**Comment by heejaechang** [comment]
@AlekseyTs I did what you asked.

**Comment by heejaechang** [comment]
ping?

**Comment by heejaechang** [comment]
@dotnet/roslyn-infrastructure tests are not finishing.

File: src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb [hunk 2]
```diff
@@ -124,5 +120,23 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         Public Shared Function ShortNameWithTypeArgsAndContainingTypes(symbol As Symbol) As FormattedSymbol
             Return New FormattedSymbol(symbol, ShortWithTypeArgsAndContainingTypesFormat)
         End Function
+
+        Public Shared Function DefaultErrorFormat(symbol As Symbol) As FormattedSymbol
+            Return New FormattedSymbol(symbol, SymbolDisplayFormat.VisualBasicErrorMessageFormat)
+        End Function
+
+        Public Shared Function DefaultErrorFormatIfErrorSymbol(arg As Object) As Object
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb]: Why is there "ErrorSymbol" in the name, I don't see any special handling for error symbols, what I see is a special handling for special types. Perhaps the name should be "DefaultErrorFormatIfSpecialType"?

File: src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb [hunk 3]
```diff
@@ -124,5 +120,23 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
+
+            Dim symbol = TryCast(arg, Symbol)
+            If symbol Is Nothing Then
+                Return arg
+            End If
+
+            If TypeOf symbol Is ITypeSymbol AndAlso DirectCast(symbol, ITypeSymbol).SpecialType <> SpecialType.None Then
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb]: In compiler we prefer Kind checks to type checks. Also, compilers do not operate in terms of symbol interfaces.

File: src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb [hunk 4]
```diff
@@ -124,5 +120,23 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
             Return New FormattedSymbol(symbol, ShortWithTypeArgsAndContainingTypesFormat)
         End Function
+
+        Public Shared Function DefaultErrorFormat(symbol As Symbol) As FormattedSymbol
+            Return New FormattedSymbol(symbol, SymbolDisplayFormat.VisualBasicErrorMessageFormat)
+        End Function
+
+        Public Shared Function DefaultErrorFormatIfErrorSymbol(arg As Object) As Object
+
+            Dim symbol = TryCast(arg, Symbol)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb]: What is the point to cast to Symbol first if we are interested in types only? Cast directly to TypeSymbol?

File: src/Compilers/VisualBasic/Portable/Emit/PEModuleBuilder.vb [hunk 5]
```diff
@@ -443,23 +444,32 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Emit
...
                                         ' Forwarded type conflicts with exported type
-                                        context.Diagnostics.Add(New VBDiagnostic(ErrorFactory.ErrorInfo(ERRID.ERR_ForwardedTypeConflictsWithExportedType,
-                                                                                                    exportedType, exportedType.ContainingAssembly,
-                                                                                                    contender, contender.ContainingModule),
-                                            NoLocation.Singleton))
+                                        context.Diagnostics.Add(New VBDiagnostic(ErrorFactory.ErrorInfo(
+                                                                                    ERRID.ERR_ForwardedTypeConflictsWithExportedType,
+                                                                                    CustomSymbolDisplayFormatter.DefaultErrorFormat(exportedType),
+                                                                                    CustomSymbolDisplayFormatter.DefaultErrorFormat(exportedType.ContainingAssembly),
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Emit/PEModuleBuilder.vb]: It doesn't look like we need to do this for AssemblySymbols, ConvertSymbolToString always formats them the same way regardless of the error number. The goal, as I understand it, is to get rid of special cases, which are based on error numbers.

File: src/Compilers/VisualBasic/Portable/Emit/PEModuleBuilder.vb [hunk 6]
```diff
@@ -443,23 +444,32 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Emit
...
                                     Else
                                         ' Forwarded type conflicts with another forwarded type
-                                        context.Diagnostics.Add(New VBDiagnostic(ErrorFactory.ErrorInfo(ERRID.ERR_ForwardedTypesConflict,
-                                                                                                    exportedType, exportedType.ContainingAssembly,
-                                                                                                    contender, contender.ContainingAssembly),
-                                            NoLocation.Singleton))
+                                        context.Diagnostics.Add(New VBDiagnostic(ErrorFactory.ErrorInfo(
+                                                                                    ERRID.ERR_ForwardedTypesConflict,
+                                                                                    CustomSymbolDisplayFormatter.DefaultErrorFormat(exportedType),
+                                                                                    CustomSymbolDisplayFormatter.DefaultErrorFormat(exportedType.ContainingAssembly),
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Emit/PEModuleBuilder.vb]: Formatting assembly here as well.

File: src/Compilers/VisualBasic/Portable/Emit/PEModuleBuilder.vb [hunk 7]
```diff
@@ -443,23 +444,32 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Emit
...
+                                                                                    CustomSymbolDisplayFormatter.DefaultErrorFormat(contender),
+                                                                                    CustomSymbolDisplayFormatter.DefaultErrorFormat(contender.ContainingAssembly)),
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Emit/PEModuleBuilder.vb]: Formatting assembly here as well.

File: src/Compilers/VisualBasic/Portable/Symbols/EventSymbol.vb [hunk 8]
```diff
@@ -207,7 +207,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
                         ' TODO: Perhaps the error wording could be changed a bit to say "type of event..." ?
                         '
                         ' Reference required to assembly '{0}' containing the definition for event '{1}'. Add one to your project.
-                        errorInfo = ErrorFactory.ErrorInfo(ERRID.ERR_UnreferencedAssemblyEvent3, errorInfo.Arguments(0), Me)
+                        errorInfo = ErrorFactory.ErrorInfo(ERRID.ERR_UnreferencedAssemblyEvent3,
+                                                           CustomSymbolDisplayFormatter.DefaultErrorFormatIfErrorSymbol(errorInfo.Arguments(0)),
+                                                           CustomSymbolDisplayFormatter.DefaultErrorFormatIfErrorSymbol(Me))
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/EventSymbol.vb]: Why is this change needed, it doesn't look like ERR_UnreferencedAssemblyEvent3 was specially handled in ConvertSymbolToString?

File: src/Compilers/VisualBasic/Portable/Symbols/MissingMetadataTypeSymbol.vb [hunk 9]
```diff
@@ -50,7 +50,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
                 Dim containingAssembly As AssemblySymbol = Me.ContainingAssembly
 
                 If containingAssembly.IsMissing Then
-                    Return ErrorFactory.ErrorInfo(ERRID.ERR_UnreferencedAssembly3, containingAssembly.Identity, Me)
+                    Return ErrorFactory.ErrorInfo(ERRID.ERR_UnreferencedAssembly3,
+                                                  CustomSymbolDisplayFormatter.DefaultErrorFormatIfErrorSymbol(containingAssembly.Identity),
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/MissingMetadataTypeSymbol.vb]: Identity is not a symbol, there is no good reason to call DefaultErrorFormatIfErrorSymbol here.

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceNamedTypeSymbol.vb [hunk 10]
```diff
@@ -545,6 +542,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
 
                         Else
                             ' Neither of types is embedded.
+                            Dim _6thArg As Object = Me.ContainingSymbol.ToErrorMessageArgument(ERRID.ERR_TypeConflict6)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceNamedTypeSymbol.vb]: I think we should inline this local, there is no good reason to create the argument before we decided that we need it.

File: src/Compilers/VisualBasic/Portable/Symbols/Symbol.vb [hunk 11]
```diff
@@ -877,7 +877,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                 Select Case errorInfo.Code
                     Case ERRID.ERR_UnreferencedAssemblyBase3,
                          ERRID.ERR_UnreferencedAssemblyImplements3
-                        errorInfo = ErrorFactory.ErrorInfo(ERRID.ERR_UnreferencedAssembly3, errorInfo.Arguments(0), errorInfo.Arguments(1))
+                        errorInfo = ErrorFactory.ErrorInfo(ERRID.ERR_UnreferencedAssembly3,
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/Symbol.vb]: This looks like a dead code. I do not see any place in the compiler where it can create DiagnosticInfo for ERRID.ERR_UnreferencedAssemblyBase3/ERRID.ERR_UnreferencedAssemblyImplements3. I think we should comment out the enum members (to avoid reuse of the numbers), remove this code and remove corresponding error messages from .RESX file.

File: src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb]: Taking all suggestions into account, there is only one meaningful call site for this function, it is in MissingMetadataTypeSymbol, which is a TypeSymbol. I don't think it is worth introducing a special helper for it. We should simply inline the logic for special types there.

**Comment by AlekseyTs** [comment]
Done with review pass.

File: src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb (hunk 2 above)
Comment (heejaechang) [inline src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb]: it is used in more than missingMetadataTypeSymbol so, I dont think I can inline there.

File: src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb (hunk 3 above)
Comment (heejaechang) [inline src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb]: do you mean I check for all type kind?

File: src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb (hunk 4 above)
Comment (heejaechang) [inline src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb]: given argument can be not symbol like FormatedSymbol.

File: src/Compilers/VisualBasic/Portable/Emit/PEModuleBuilder.vb (hunk 5 above)
Comment (heejaechang) [inline src/Compilers/VisualBasic/Portable/Emit/PEModuleBuilder.vb]: won't this make logic to spread too much?

File: src/Compilers/VisualBasic/Portable/Emit/PEModuleBuilder.vb (hunk 6 above)
Comment (heejaechang) [inline src/Compilers/VisualBasic/Portable/Emit/PEModuleBuilder.vb]: done

File: src/Compilers/VisualBasic/Portable/Emit/PEModuleBuilder.vb (hunk 7 above)
Comment (heejaechang) [inline src/Compilers/VisualBasic/Portable/Emit/PEModuleBuilder.vb]: done

File: src/Compilers/VisualBasic/Portable/Symbols/EventSymbol.vb (hunk 8 above)
Comment (heejaechang) [inline src/Compilers/VisualBasic/Portable/Symbols/EventSymbol.vb]: ah good catch. errid in case statement and actual error id wasn't match.

File: src/Compilers/VisualBasic/Portable/Symbols/MissingMetadataTypeSymbol.vb (hunk 9 above)
Comment (heejaechang) [inline src/Compilers/VisualBasic/Portable/Symbols/MissingMetadataTypeSymbol.vb]: done

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceNamedTypeSymbol.vb (hunk 10 above)
Comment (heejaechang) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceNamedTypeSymbol.vb]: done

File: src/Compilers/VisualBasic/Portable/Symbols/Symbol.vb (hunk 11 above)
Comment (heejaechang) [inline src/Compilers/VisualBasic/Portable/Symbols/Symbol.vb]: done

**Comment by heejaechang** [comment]
@dotnet/roslyn-infrastructure where can I see test artifacts to find out why test has failed?

File: src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb]: There was a reason why I said: "Taking all suggestions into account". There is only one call site for this method right now.

File: src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb (hunk 4 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb]: But it looks like we are interested in TypeSymbols only, why not check for TypeSymbol from the start?

File: src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb [hunk 12]
```diff
@@ -124,5 +122,23 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
             Return New FormattedSymbol(symbol, ShortWithTypeArgsAndContainingTypesFormat)
         End Function
+
+        Public Shared Function DefaultErrorFormat(symbol As Symbol) As FormattedSymbol
+            Return New FormattedSymbol(symbol, SymbolDisplayFormat.VisualBasicErrorMessageFormat)
+        End Function
+
+        Public Shared Function DefaultErrorFormatIfSpecialType(arg As Object) As Object
+
+            Dim symbol = TryCast(arg, Symbol)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb]: I still don't see a good reason to check for Symbol first and then check to TypeSymbol. We should simply go for TypeSymbol here.

File: src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb [hunk 13]
```diff
@@ -124,5 +122,23 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         Public Shared Function ShortNameWithTypeArgsAndContainingTypes(symbol As Symbol) As FormattedSymbol
             Return New FormattedSymbol(symbol, ShortWithTypeArgsAndContainingTypesFormat)
         End Function
+
+        Public Shared Function DefaultErrorFormat(symbol As Symbol) As FormattedSymbol
+            Return New FormattedSymbol(symbol, SymbolDisplayFormat.VisualBasicErrorMessageFormat)
+        End Function
+
+        Public Shared Function DefaultErrorFormatIfSpecialType(arg As Object) As Object
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb]: We should inline this helper, there is only one call site for it in MissingMetadataTypeSymbol and the logic can be much simpler because we know we are dealing with TypeSymbol.

File: src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb (hunk 12 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb]: Of course, if we inline DefaultErrorFormatIfSpecialType, this logic becomes obsolete.

**Comment by heejaechang** [comment]
per this email "This is on the tst machines.  They are expected to fail.  Last I checked, Tanner was looking into them."

I am re-running the test.

**Comment by heejaechang** [comment]
@dotnet-bot retest  prtest/win/tst/eta please

**Comment by heejaechang** [comment]
@dotnet-bot retest this please

**Comment by heejaechang** [comment]
@dotnet/roslyn-infrastructure this is happening again - https://github.com/dotnet/roslyn/issues/8077

**Comment by heejaechang** [comment]
@dotnet-bot retest prtest/win/dbg/unit32 please

...

this is known issue.

**Comment by heejaechang** [comment]
@dotnet/roslyn-infrastructure retest seems not working

**Comment by heejaechang** [comment]
retest prtest/win/dbg/unit32 please

**Comment by heejaechang** [comment]
retest win unit32 please

**Comment by heejaechang** [comment]
retest unit please

**Comment by heejaechang** [comment]
retest  prtest/win/dbg/unit32 please

**Comment by heejaechang** [comment]
retest prtest/win/dbg/unit32 please

I dont believe failure is related to my change. opened an issue.

**Comment by heejaechang** [comment]
retest prtest/win/dbg/unit32 please

**Comment by heejaechang** [comment]
it looks like I re-ran didn't pick up the fix checked in by tmas. will re-run

**Comment by heejaechang** [comment]
retest prtest/win/dbg/unit32 please

**Comment by heejaechang** [comment]
known issue that closed as resolved happened again. re-opened the issue.

**Comment by heejaechang** [comment]
retest prtest/win/dbg/unit32 please

**Comment by heejaechang** [comment]
@davkean is the issue resolved, can I re-run the test?

**Comment by davkean** [comment]
@dotnet-bot retest prtest/win/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/3570/
// Retest reason: #8498

**Comment by heejaechang** [comment]
@dotnet/roslyn-infrastructure looks like an IO failure happened for  prtest/determinism
http://dotnet-ci.cloudapp.net/job/roslyn_prtest_determinism/50/

**Comment by heejaechang** [comment]
retest prtest/determinism please

**Comment by heejaechang** [comment]
retest prtest/determinism please

**Comment by heejaechang** [comment]
both tests failures are known, and told to ignore. so I am checking in.

File: src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb (hunk 12 above)
Comment (heejaechang) [inline src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb]: yep, you are right, intention was bailing out right away if given arg is not symbol at all (such as FormattedSymbol). but at the end result is same if it is not typesymbol.

Comment (heejaechang) [inline src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb]: I think I wrote in other comment. there are multiple place where DefaultErrorFormatIfSpecialType is used.

File: src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb (hunk 13 above)
Comment (heejaechang) [inline src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb]: there are multiple places I am not sure why you think there is only one call site.

**Comment by heejaechang** [comment]
@dotnet/roslyn-compiler can I get another code review? I am reverting the checkin to get another sign off.

File: src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb (hunk 12 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb]: Perhaps you should show those multiple places?

File: src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb (hunk 13 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb]: Perhaps you should show those multiple places?

**Comment by AlekseyTs** [comment]
There should be at least two sign-offs from Compiler team. At the moment there are **none**.

**Comment by heejaechang** [comment]
@AlekseyTs yea, I think I misinterpreted "Done with review pass." this.

File: src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb (hunk 12 above)
Comment (heejaechang) [inline src/Compilers/VisualBasic/Portable/SymbolDisplay/CustomSymbolDisplayFormatter.vb]: ya, you are right again, after my last PR, all other places except missingmetadatatypesymbol are removed.

**Comment by CyrusNajmabadi** [comment]
> fix devdiv bug #118143

Can you be more specific?  What is the bug.  What is getting fixed?  How are things being fixed?

**Comment by heejaechang** [comment]
@CyrusNajmabadi unfortunately I can't link devdiv bug here. you need to go bung or devdiv vso to see  the bug.

bug is basically, remove any special case from MessageProvider.ConvertSymbolToString and inline those case at callsite with FormatedSymbol.

**Comment by jaredpar** [comment]
@heejaechang  why not?  

https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_workitems?_a=edit&id=118143

**Comment by heejaechang** [comment]
@jaredpar I thought we dont share any internal stuff such as link to index, fileshare, bung, devdiv bug and etc.

**Comment by jaredpar** [comment]
@heejaechang we track all bugs by link now in our `WorkItem` attributes.
